components/script/dom/node.rs

```diff
@@ -103,6 +109,9 @@ pub struct Node {
     /// The parent of this node.
     parent_node: MutNullableDom<Node>,
 
+    /// The parent of this node, ignoring shadow roots.
+    composed_parent_node: MutNullableDom<Node>,
```

Comment:
Why adding the composed parent node here? Seems unfortunate to grow `Node` just for this, when computing using a is straight-forward and would simplify the patch a lot.